_posts/2018-09-05-Data-Continuity-Service-for-DynamoDB.md

```diff
@@ -0,0 +1,570 @@
+
+Or use the CloudFormation snippet below:
+
+```json
+{
+  "DynamoDbStreamToS3LambdaFunction": {
+    "Properties": {
+      "Code": {
+        "ZipFile": "exports.backup = function (event, context, callback){ console.log('Hello'); callback(null); }"
+      },
+      "Environment": {
+        "Variables": {
+          "BUCKET": "source-account-dynamodb-backup",
+          "TABLE_NAME": "users"
+        }
+      },
+      "Handler": "index.backup",
+      "MemorySize": 128,
+      "Role": {
+        "Fn::GetAtt": [
+          "DynamoDbStreamToS3LambdaRole",
+          "Arn"
+        ]
+      },
+      "Runtime": "nodejs6.10",
+      "Timeout": 300
+    },
+    "Type": "AWS::Lambda::Function"
+  }
+}
+```
+
+After the function is created we need to attach the DynamoDB Stream to our Lambda function so the function can listen and work whenever there is an event posted:
+
+![configure lambda 1](/assets/posts/2018-09-05-Data-Continuity-Service-for-DynamoDB/configure-lambda-trigger-1.png)
+
+Resulting in:
+
+![configure lambda 2](/assets/posts/2018-09-05-Data-Continuity-Service-for-DynamoDB/configure-lambda-trigger-2.png)
+
+Or update your CloudFormation template with the content below below:
+```json
+{
+  "DynamoDbStreamToS3LambdaEventSourceMapping": {
+    "Properties": {
+      "EventSourceArn": "arn:aws:dynamodb:<YOUR_AWS_REGION>:<YOUR_AWS_ACCOUNT_ID>:table/users/stream/<STREAM_ID>",
+      "FunctionName": {
+        "Ref": "DynamoDbStreamToS3LambdaFunction"
+      },
+      "StartingPosition": "LATEST"
+    },
+    "Type": "AWS::Lambda::EventSourceMapping"
+  }
+}
+```
+
+Deployed? Done? Good. Everything on the Lambda side is ready apart from the function itself. Luckily the function itself is rather simple:
+
+```javascript
+'use strict';
+
+const aws = require('aws-sdk');
+
+const s3 = new aws.S3({ region: '<YOUR AWS_REGION>' });
+
+module.exports.backup = (event, context, callback) => {
+  const records = event.Records;
+
+  Promise.all(records.map((record) => {
+    const keysList = Object.keys(record.dynamodb.Keys).map((key) => {
+      const keyDefinition = record.dynamodb.Keys[key];
+      const type = Object.keys(keyDefinition)[0];
+      const value = keyDefinition[type];
+      return value;
+    });
+
+    const keysString = keysList[0];
+    let image = {}
+    if(typeof record.dynamodb.NewImage !== 'undefined') {
+      image = aws.DynamoDB.Converter.output({ M: record.dynamodb.NewImage });
+    }
+
+    return s3.putObject({
+      Bucket: process.env.BUCKET,
+      Key: `${process.env.TABLE_NAME}/${keysString}/image.json`,
+      Body: JSON.stringify(image),
+      StorageClass: "STANDARD_IA"
+    }).promise()
+      .then((response) => {
+        console.log(`${keysString} snapshot done`, response);
+      })
+      .catch((err) => {
+        console.error('Error', err);
+      });
+  }))
+    .then(v => callback(null, v), callback);
+};
+```
```

Comment:
Maybe define the Lambda language explicitly somewhere, (NodeJS 6.10), because it's only in a screenshot now.